server/src/ArmUtils.cc

```diff
@@ -0,0 +1,250 @@
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with Hatohol. If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include <SmartTime.h>
+#include "ArmUtils.h"
+#include "UnifiedDataStore.h"
+#include "ThreadLocalDBCache.h"
+
+using namespace std;
+using namespace mlpl;
+
+static const char *SERVER_SELF_MONITORING_SUFFIX = "_SELF";
+
+struct ArmUtils::Impl
+{
+	ArmUtils                   *obj;
+	const MonitoringServerInfo &serverInfo;
+	ArmTrigger                 *armTriggers;
+	const size_t                numArmTriggers;
+
+	Impl(ArmUtils *_obj, const MonitoringServerInfo &_serverInfo,
+	     ArmTrigger *_armTriggers, const size_t &_numArmTriggers)
+	: obj(_obj),
+	  serverInfo(_serverInfo),
+	  armTriggers(_armTriggers),
+	  numArmTriggers(_numArmTriggers)
+	{
+		initializeArmTriggers();
+	}
+
+	void initializeArmTriggers(void)
+	{
+		for (size_t i = 0; i < numArmTriggers; i++)
+			armTriggers[i].status = TRIGGER_STATUS_ALL;
+	}
+
+	bool isCollectingOK(const size_t &idx)
+	{
+		return (idx == numArmTriggers + 1);
+	}
+
+	bool isUsed(const size_t &idx)
+	{
+		return (armTriggers[idx].status != TRIGGER_STATUS_ALL);
+	}
+
+	void updateTriggerStatus(
+	  const size_t &triggerIdx, const TriggerStatusType &status)
+	{
+		UpdateTriggerArg arg = {triggerIdx, status};
+
+		if (isCollectingOK(triggerIdx))
+			updateTriggersForCollectingOK(arg);
+		else
+			updateTriggersForCollectingFailure(arg);
+
+		if (!arg.triggerInfoList.empty()) {
+			ThreadLocalDBCache cache;
+			cache.getMonitoring().addTriggerInfoList(
+			  arg.triggerInfoList);
+		}
+		if (!arg.eventInfoList.empty()) {
+			UnifiedDataStore *dataStore =
+			  UnifiedDataStore::getInstance();
+			dataStore->addEventList(arg.eventInfoList);
+		}
+	}
+
+private:
+	struct UpdateTriggerArg {
+		const size_t            triggerIdx;
+		const TriggerStatusType status;
+		TriggerInfoList         triggerInfoList;
+		EventInfoList           eventInfoList;
+	};
+
+	void updateTriggersForCollectingOK(UpdateTriggerArg &arg)
+	{
+		// Since collecting has successed, all the self triggers
+		// should be OK.
+		for (size_t i = 0; i < numArmTriggers; i++) {
+			if (!isUsed(i))
+				continue;
+			setTriggerToStatusOK(arg, armTriggers[i]);
+		}
+	}
+
+	void updateTriggersForCollectingFailure(UpdateTriggerArg &arg)
+	{
+		if (!isUsed(arg.triggerIdx)) {
+			MLPL_WARN("Try to update the unused trigger: %zd\n",
+			          arg.triggerIdx);
+			return;
+		}
+
+		ArmTrigger &armTrigger = armTriggers[arg.triggerIdx];
+		if (armTrigger.status == arg.status)
+			return;
+
+		// update the target trigger
+		armTrigger.status = arg.status;
+		obj->createTrigger(armTrigger, arg.triggerInfoList);
+		if (armTrigger.status != TRIGGER_STATUS_UNKNOWN)
+			obj->createEvent(armTrigger, arg.eventInfoList);
+
+		updateLowerTriggers(arg);
+		updateUpperTriggers(arg);
+	}
+
+	void updateLowerTriggers(UpdateTriggerArg &arg)
+	{
+		for (size_t i = arg.triggerIdx + 1; i < numArmTriggers; i++) {
```

Comment:
The range of the "i" seems same with updateUpperTriggers().
Is it correct?

I think the correct code for it is:
```c++
for (size_t i = 0; i < arg.triggerIdx; i++) {
```